src/FakeItEasy/Core/SequentialCallContext.cs

```diff
@@ -2,48 +2,51 @@ namespace FakeItEasy.Core
 {
     using System;
     using System.Collections.Generic;
-    using System.Globalization;
-    using System.IO;
+    using System.Linq;
 
-    internal class OrderedFakeAsserter : IFakeAsserter
+    internal class SequentialCallContext : ISequentialCallContext
     {
-        private readonly IEnumerable<IFakeObjectCall> originalCallList;
         private readonly CallWriter callWriter;
-        private readonly Queue<IFakeObjectCall> calls;
-        private readonly List<AssertedCall> assertedCalls = new List<AssertedCall>();
+        private readonly HashSet<FakeManager> fakeManagers;
+        private readonly List<AssertedCall> assertedCalls;
+        private int currentSequenceNumber;
 
-        public OrderedFakeAsserter(IEnumerable<IFakeObjectCall> calls, CallWriter callWriter)
+        public SequentialCallContext(CallWriter callWriter)
         {
-            Guard.AgainstNull(calls, "calls");
-            Guard.AgainstNull(callWriter, "callWriter");
-
-            this.originalCallList = calls;
-            this.calls = new Queue<IFakeObjectCall>(calls);
             this.callWriter = callWriter;
+            this.currentSequenceNumber = -1;
+            this.fakeManagers = new HashSet<FakeManager>();
+            this.assertedCalls = new List<AssertedCall>();
         }
 
-        public virtual void AssertWasCalled(
-            Func<IFakeObjectCall, bool> callPredicate, string callDescription, Func<int, bool> repeatPredicate, string repeatDescription)
+        public void CheckNextCall(FakeManager fakeManager, Func<IFakeObjectCall, bool> callPredicate, string callDescription, Repeated repeatConstraint)
         {
-            this.assertedCalls.Add(new AssertedCall { CallDescription = callDescription, RepeatDescription = repeatDescription });
-            var matchedCallCount = 0;
-            while (!repeatPredicate(matchedCallCount))
+            this.fakeManagers.Add(fakeManager);
+            this.assertedCalls.Add(new AssertedCall { CallDescription = callDescription, RepeatDescription = repeatConstraint.ToString() });
+            var allCalls = this.fakeManagers.SelectMany(f => f.GetRecordedCalls().Cast<IFakeObjectCall>()).OrderBy(c => c.SequenceNumber).ToList();
+
+            int matchedCallCount = 0;
+            foreach (var currentCall in allCalls.SkipWhile(c => c.SequenceNumber <= this.currentSequenceNumber))
             {
-                if (this.calls.Count == 0)
+                if (repeatConstraint.Matches(matchedCallCount))
                 {
-                    ThrowExceptionWhenAssertionFailed(this.assertedCalls, this.callWriter, this.originalCallList);
+                    break;

```

Comment:
Well, in this case it doesn't really matter, since the previous statement throws anyway, but I agree that `return` would be clearer. Seeing this again, an even better approach would be to make `ThrowExceptionWhenAssertionFailed` return an exception (rather than throw it), and throw that exception explicitly from `CheckNextCall`.